File: Prolog/Prolog_Lists/problem5.pl
%% 1.05 (*) Reverse a list.

% reversed(+Xs, ?RXs)
reversed(Xs, RXs) :- reversed(Xs, [], RXs).

reversed([], Acc, Acc).
reversed([X|Xs], Acc, RXs) :- reversed(Xs, [X|Acc], RXs).

File: Haskell/problem5.hs
{- (*) Reverse a list. -}

reversed :: [a] -> [a]
reversed = reversed_ []

reversed_ :: [a] -> [a] -> [a]
reversed_ acc [] = acc
reversed_ acc (x:xs) = reversed_ (x:acc) xs

reversed' :: [a] -> [a]
reversed' = foldl (\acc x -> x:acc) []

reversed'' :: [a] -> [a]
reversed'' = foldl (flip (:)) []
